lib/hsa/mcwamp_hsa.cpp

```diff
@@ -271,6 +271,143 @@ static unsigned extractBits(unsigned v, unsigned pos, unsigned w)
+    {
+        using namespace std;
+
+        const auto it = find_if(
+            reader.sections.begin(), reader.sections.end(), move(p));
+
+        return it != reader.sections.end() ? *it : nullptr;
+    }
+
+    inline
+    std::vector<std::string> copy_names_of_undefined_symbols(
+        const ELFIO::symbol_section_accessor& section)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        vector<string> r;
+
+        for (auto i = 0u; i != section.get_symbols_num(); ++i) {
+            // TODO: this is boyscout code, caching the temporaries
+            //       may be of worth.
+            string name;
+            Elf64_Addr value = 0;
+            Elf_Xword size = 0;
+            Elf_Half sect_idx = 0;
+            uint8_t bind = 0;
+            uint8_t type = 0;
+            uint8_t other = 0;
+
+            section.get_symbol(
+                i, name, value, size, bind, type, sect_idx, other);
+
+            if (sect_idx == SHN_UNDEF && !name.empty()) {
+                r.push_back(std::move(name));
+            }
+        }
+
+        return r;
+    }
+
+    inline
+    std::pair<ELFIO::Elf64_Addr, ELFIO::Elf_Xword> find_symbol_address(
+        const ELFIO::symbol_section_accessor& section,
+        const std::string& symbol_name)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        static constexpr pair<Elf64_Addr, Elf_Xword> r{0, 0};
+
+        for (auto i = 0u; i != section.get_symbols_num(); ++i) {
+            // TODO: this is boyscout code, caching the temporaries
+            //       may be of worth.
+            string name;
+            Elf64_Addr value = 0;
+            Elf_Xword size = 0;
+            Elf_Half sect_idx = 0;
+            uint8_t bind = 0;
+            uint8_t type = 0;
+            uint8_t other = 0;
+
+            section.get_symbol(
+                i, name, value, size, bind, type, sect_idx, other);
+
+            if (name == symbol_name) return make_pair(value, size);
+        }
+
+        return r;
+    }
+
+    inline
+    void associate_code_object_symbols_with_host_allocation(
+        const ELFIO::elfio& reader,
+        const ELFIO::elfio& self_reader,
+        ELFIO::section* code_object_dynsym,
+        ELFIO::section* process_symtab,
+        hsa_agent_t agent,
+        hsa_executable_t executable)
+    {
+        using namespace ELFIO;
+        using namespace std;
+
+        if (!code_object_dynsym || !process_symtab) return;
+
+        const auto undefined_symbols = copy_names_of_undefined_symbols(
+            symbol_section_accessor{reader, code_object_dynsym});
+
+        for (auto&& x : undefined_symbols) {
+            const auto tmp = find_symbol_address(
+                symbol_section_accessor{self_reader, process_symtab}, x);
+
+            assert(tmp.first);
+
+            void* p = nullptr;
+            hsa_amd_memory_lock(
+                reinterpret_cast<void*>(tmp.first), tmp.second, &agent, 1, &p);
```

Comment:
We could, but this would require either using a slightly more complex data structure or O(n) searching on each lock. If the mapping is bijective, this appears unnecessary, for there can be no scenario in which multiple different symbols try to lock the same loader allocated variable in the executable's image. If we were to have multiple code-objects containing code which references the same global, and which are loaded during the same execution, this may be an optimisation worth pursuing. At the moment we do not support it, but a scenario for it would be both the executable and shared object(s) referencing a global that is defined in a shared object.

Reply:
Any reason why we don't support multiple objects referencing the same global?  It seems to be that your algorithm should just work and we just need verify whether ROCR is able to handle double locking/unlocking the same global.  If ROCR couldn't handle it properly than hcc RT would need to manage the memory lock/unlock more carefully.